lms/services/blackboard_api.py

```diff
@@ -0,0 +1,60 @@
+from lms.validation.authentication import OAuthTokenResponseSchema
+
+
+class BlackboardAPIClient:
+    def __init__(
+        self,
+        blackboard_host,
+        client_id,
+        client_secret,
+        redirect_uri,
+        http_service,
+        oauth2_token_service,
+    ):  # pylint:disable=too-many-arguments
+        self.blackboard_host = blackboard_host
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.redirect_uri = redirect_uri
+
+        self._http_service = http_service
+        self._oauth2_token_service = oauth2_token_service
+
+    def get_token(self, authorization_code):
+        """
+        Get an access token from Blackboard and save it in the DB.
+
+        :raise services.HTTPError: if something goes wrong with the access
+            token request to Blackboard
+        """
+        # Send a request to Blackboard to get an access token.
+        response = self._http_service.post(
+            f"https://{self.blackboard_host}/learn/api/public/v1/oauth2/token",
+            data={
+                "grant_type": "authorization_code",
+                "redirect_uri": self.redirect_uri,
+                "code": authorization_code,
+            },
+            auth=(self.client_id, self.client_secret),
+            schema=OAuthTokenResponseSchema,
```

Comment:
We've mentioned this a few times, but I think putting the responsibility for validating the data (the format of which is a very high level concept) into the lowest HTTP layer is a bit strange and causes some problems leading to the clarity vs. isolation design problems in the unit tests.

Reply:
I'm not sure I agree. I think "send this request and validate the response using this schema" is a pretty logical unit, and something that's frequently needed.

`HTTPService` isn't responsible for the format of the data, it's just responsible for applying the given schema.

`response = http_service.post(..., schema=FooSchema)` seems clear to me. It makes it clear and explicit that `FooSchema` will have been applied to `response` and that the method can expect `response.validated_data` to contain the expected fields as defined by `FooSchema`, or a correctly-formatted exception to have been raised.

It's pretty analogous to guard schemas for views:

```python
@view_config(..., schema=FooSchema)
def foo_view(request):
    ...
```

The view can assume that `request.parsed_params` will contain the valid params as defined by `FooSchema` or an exception will have been raised.